Mixpanel/AutomaticEvents.m

```diff
@@ -0,0 +1,158 @@
+//
+//  AutomaticEvents.m
+//  Mixpanel
+//
+//  Created by Yarden Eitan on 4/18/17.
+//  Copyright © 2017 Mixpanel. All rights reserved.
+//
+
+#import "AutomaticEvents.h"
+#import "MPSwizzler.h"
+#import <objc/runtime.h>
+
+@implementation AutomaticEvents {
+    NSMutableDictionary *awaitingTransactions;
+    NSUserDefaults *defaults;
+    NSTimeInterval sessionLength;
+    NSTimeInterval sessionStartTime;
+    MixpanelPeople *people;
+}
+
+static NSTimeInterval _appStartTime;
++ (NSTimeInterval)appStartTime { return _appStartTime; }
++ (void)setAppStartTime:(NSTimeInterval)appStartTime { _appStartTime = appStartTime; }
+
+__attribute__((constructor))
+static void initialize_appStartTime() {
+    if (AutomaticEvents.appStartTime == 0) {
+        AutomaticEvents.appStartTime = [[NSDate date] timeIntervalSince1970];
+    }
+}
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        awaitingTransactions = [[NSMutableDictionary alloc] init];
+        defaults = [[NSUserDefaults alloc] initWithSuiteName:@"Mixpanel"];
+        sessionLength = 0;
+        sessionStartTime = 0;
+        self.minimumSessionDuration = 10000;
+        self.maximumSessionDuration = UINT64_MAX;
+    }
+    return self;
+}
+
+- (void)initializeEvents:(MixpanelPeople *)peopleInstance {
+    people = peopleInstance;
+    NSString *firstOpenKey = @"MPFirstOpen";
+    if (defaults != nil && ![defaults boolForKey:firstOpenKey]) {
+        if (![self isExistingUser]) {
+            [self.delegate track:@"$ae_first_open" properties:nil];
+            [people setOnce:@{@"$ae_first_app_open_date": [NSDate date]}];
+        }
+        [defaults setBool:TRUE forKey:firstOpenKey];
+        [defaults synchronize];
+    }
+
+    NSDictionary* infoDict = [NSBundle mainBundle].infoDictionary;
+    if (defaults != nil && infoDict != nil) {
+        NSString* appVersionKey = @"MPAppVersion";
+        NSString* appVersionValue = infoDict[@"CFBundleShortVersionString"];
+        NSString* savedVersionValue = [defaults stringForKey:appVersionKey];
+        if (appVersionValue != nil && savedVersionValue != nil && ![appVersionValue isEqualToString:savedVersionValue]) {
```

Comment:
yup, what i mean is that we only check that the stored string is different from the new short bundle version. we should make sure that the version is actually a new one (i.e. X.X.X > X.X.Y). a person could downgrade or during development you might change the versions and it would send App Update event when it's not an update!